### ISSUE TITLE
About trainers fn：SFTTrainer_head.fit and PPOTrainer.kl_penalized_reward

### ISSUE THREAD

**Issue by Clingmoon** [Mowenyii/PAE#8]
Hello,
Thank you for your open source.
Is there a bit of a bug in the order of gradient clipping and backpropagation when calculating the loss in train_stage_1?
Besides that, it seems that the assignment to `estimated_kl` has been forgotten.
```python
#  SFTTrainer_head.fit
if self.grad_clip != 0.0:
    torch.nn.utils.clip_grad_norm_(opt_model.parameters(), self.grad_clip)

scaler.scale(loss).backward()
scaler.step(self.optimizer)
scaler.update()
self.optimizer.zero_grad(set_to_none=True)
lossf = loss.item()
#################################
# PPOTrainer.kl_penalized_reward
    def kl_penalized_reward(
        self,
        reward: torch.Tensor,
        log_prob_rl: torch.Tensor,
        log_prob_sft: torch.Tensor,
        action_mask: torch.Tensor = None,
    ) -> Tuple[Tensor, Tensor]:
        # log(π_RL(y|x) / π_SFL(y|x)) = log(π_RL(y|x)) - log(π_SFL(y|x))
        ratio = log_prob_rl - log_prob_sft
        # k3 in http://joschu.net/blog/kl-approx.html
        estimated_kl = (torch.exp(ratio) - 1) - ratio
        if action_mask: #################################################### please pay attention here
            estimated_kl = estimated_kl * action_mask
            estimated_kl.sum(dim=1) / action_mask.sum(dim=1)
        estimated_kl = estimated_kl.mean(
            dim=1, keepdim=True)  # estimated_kl -> (B, 1)
        return reward - self.cfg.kl_beta * estimated_kl, estimated_kl
```
Is it possible to modify it to:
```python
# SFTTrainer_head.fit
scaler.scale(loss).backward() 
scaler.unscale_(self.optimizer) 
if self.grad_clip != 0.0:
    torch.nn.utils.clip_grad_norm_(opt_model.parameters(), self.grad_clip)
scaler.step(self.optimizer) 
scaler.update() 
self.optimizer.zero_grad(set_to_none=True)  
############################
# PPOTrainer.kl_penalized_reward
if action_mask is not None:
    estimated_kl = (estimated_kl * action_mask).sum(dim=1) / action_mask.sum(dim=1)
else:
    estimated_kl = estimated_kl.mean(dim=1)
estimated_kl = estimated_kl.unsqueeze(-1)  # (B,1)
```
Looking forward to your reply.

**Comment by Mowenyii** [comment]
Hello,

Thank you very much for your careful review and suggestions!

You are right — this code strongly depends on the current implementations at
https://github.com/ethanyanjiali/minChatGPT/blob/main/src/trainers.py#L386 and
https://github.com/ethanyanjiali/minChatGPT/blob/main/src/trainers.py#L167.

Regarding the order of gradient clipping and backpropagation in train_stage_1, and the handling of estimated_kl in kl_penalized_reward, we really appreciate you pointing them out. Your proposed modifications make a lot of sense and would improve the stability and correctness of the training process.

We will consider incorporating these changes and update the code accordingly.
Thanks again for your valuable feedback!
